### PR TITLE
Fix signature of printf-like function (pointer) types

### PR DESCRIPTION
**Pull request by nsajko**

### CODE REVIEW

**Comment by radare** [comment]
The change lgtm but i would keep the void return because we dont use it anywhere and the result is pretty incomplete right now so it wouldnt be helpful even if we use it

**Comment by nsajko** [comment]
@radare, the `int` return type is a must because the type signature has to be the same like printf (`int printf(const char * restrict format, ...);`).

**Comment by radare** [comment]
that makes windows unhappy

**Comment by radare** [comment]
screen shot 2018-09-14 at 16 03 32

**Comment by nsajko** [comment]
Wow, the Windows compiler does not support `restrict`, although it has been in C since C99.

**Comment by radare** [comment]
probably because of the restrict keyword, can you get rid of it or use a wrapped define to use alternative constructions.

File: libr/cons/cons.c [hunk 1]
```diff
@@ -928,14 +928,16 @@ R_API void r_cons_printf_list(const char *format, va_list ap) {
 	va_end (ap3);
 }
 
-R_API void r_cons_printf(const char *format, ...) {
+R_API int r_cons_printf(const char *restrict format, ...) {
```
Comment (ret2libc) [inline libr/cons/cons.c]: Why do we need `restrict`?

Comment (nsajko) [reply libr/cons/cons.c]: I said it already in my second comment on this page:

> the type signature has to be the same like printf (`int printf(const char * restrict format, ...);`).

Comment (nsajko) [reply libr/cons/cons.c]: The types have to be compatible because printf gets assigned to it.

Comment (ret2libc) [reply libr/cons/cons.c]: > I said it already in my second comment on this page:
>>    the type signature has to be the same like printf (int printf(const char * restrict format, ...);).

Sorry, I thought it was just about the `int` return type.
Two questions: 
1) I do see `int printf(const char *format, ...);` in my man page, without the `restrict`.
2) `printf` is assigned to what?

Comment (nsajko) [reply libr/cons/cons.c]: The latest standard, C11, declares printf as `int printf(const char * restrict format, ...);`. And C99 too, I guess.

printf is assigned to struct members called `cb_printf`, so they have to be the same type.

**Comment by radare** [comment]
That may depend on the libc. So casting is the only way to avoid problems. So i would remove the restrict keyword which doesnt work on windows and will fail on some compilers

**Comment by nsajko** [comment]
>That may depend on the libc.

It is in the standard.

> So casting is the only way to avoid problems.

Sorry, I do not understand this.

> So i would remove the restrict keyword which doesnt work on windows and will fail on some compilers

I already did `#define restrict` in r_types.h, effectively removing restrict on Windows.

**Comment by ret2libc** [comment]
> I already did #define restrict in r_types.h, effectively removing restrict on Windows.

@nsajko did you try with https://msdn.microsoft.com/en-us/library/5ft82fed.aspx ?

**Comment by nsajko** [comment]
> All checks have passed.

**Comment by nsajko** [comment]
> @nsajko did you try with https://msdn.microsoft.com/en-us/library/5ft82fed.aspx ?

I did not bother to do that, do you want me to?

**Comment by ret2libc** [comment]
>> @nsajko did you try with https://msdn.microsoft.com/en-us/library/5ft82fed.aspx ?
>
> I did not bother to do that, do you want me to?

As you prefer! It's ok as it is for me, I'll leave to @radare whether he wants restrict or not.

**Comment by nsajko** [comment]
@radare, ping

**Comment by Maijin** [comment]
@nsajko I reverted partially as it was causing some builds to fail on linux see - https://github.com/unixfreaxjp/dev/blob/master/r2linux-i686.md and freebsd, do you mind checking and coming back with a better fix on the `restrict`?

**Comment by Maijin** [comment]
@nsajko I reverted partially as it was causing some builds to fail on linux thanks to the report from @unixfreaxjp - https://github.com/unixfreaxjp/dev/blob/master/r2linux-i686.md and freebsd, do you mind checking and coming back with a better fix on the restrict?

**Comment by unixfreaxjp** [comment]
@nsajko don't implement restrict keywording unless you test it in several system-wide beforehand. I practically couldn't compile r2 in some of platform for it, and thanks to @radare who recognizes the restrict was causing it.

**Comment by nsajko** [comment]
This seems like a build configuration issue. @unixfreaxjp, what is your version of GCC and what is your CFLAGS? Did you pass something like '-ansi' or '-std=c90' in CFLAGS?

If not by default (because of an ancient version of GCC, or a configuration option), the build would probably work if you passed '-std=c99' or '-std=c11' or '-std=c17' or '-std=gnu99' or '-std=gnu11' or '-std=gnu17'.

Ref https://github.com/radare/radare2/issues/11346

**Comment by unixfreaxjp** [comment]
> 
> 
> This seems like a build configuration issue. @unixfreaxjp, what is your version of GCC and what is your CFLAGS? Did you pass something like '-ansi' or '-std=c90' in CFLAGS?
> 
> If not by default (because of an ancient version of GCC, or a configuration option), the build would probably work if you passed '-std=c99' or '-std=c11' or '-std=c17' or '-std=gnu99' or '-std=gnu11' or '-std=gnu17'.
> 
> Ref #11346

Again. It is NOT my issue since it was compiled smoothly and well before the **restrict** keyword is used. 

You should test it more widely and carefully check the compatibility matter **before** implementing something that can damage the compilation process. I am using FreeBSD 10 production release and Debian Jessie and Stretch for the latest update too. Not using anything fancy for compiler matter. Please check those default dev environment beforehand.

I had much pain about this matter, do not force us with something new that can ruin the current stability of radare2 for the UNIX platform. If you want to force **restrict** keywording buff I will rollback all of my used versions on all platforms.

**Comment by nsajko** [comment]
* No need to be so aggressive.

* You did not answer my questions. The error logs linked here also do not answer them.

* restrict is not anything new.

* Surely you wish for radare2 to keep receiving new development? Why so defensive?

**Comment by unixfreaxjp** [comment]
>     * No need to be so aggressive.
>

You're not the one who had impact for that restrict stuff, I am the one who has to deal with it in all of my systems.
 
>     * You did not answer my questions. The error logs linked here also do not answer them.
>

Read well. 
 
>     * restrict is not anything new.
> 

Before you coded that way it WORKED!!! Who you wanna blame then? ME? Learn to take responsibility of what you coded maybe a good start for you, specially without doing much proper tests! Whatta developer...

>     * Surely you wish for radare2 to keep receiving new development? Why so defensive?

Good development, yes. Some hanky-panky code without testing, NO!

Defensive? NO! I am pissing off.

**Comment by nsajko** [comment]
Please answer my questions.

**Comment by unixfreaxjp** [comment]
NO.

It's in the log see it!

**Comment by nsajko** [comment]
Sigh. I searched the log for CFLAGS with no results. Likewise there is no mention of GCC version.

**Comment by Maijin** [comment]
I do not see the version of GCC in the log provided either @unixfreaxjp.

**Comment by unixfreaxjp** [comment]
@Maijin You.. stay OUT of this. I really mean this. After what happened today, DONT make it worse.


@nsajko 

```asm
$ gcc --version
gcc (GCC) 4.2.1 20070831 patched [FreeBSD]
Copyright (C) 2007 Free Software Foundation, Inc.

$ gcc --version
gcc (Debian 4.9.2-10+deb8u1) 4.9.2
Copyright (C) 2014 Free Software Foundation, Inc.

]$ cc --version
FreeBSD clang version 3.4.1 (tags/RELEASE_34/dot1-final 208032) 20140512
Target: x86_64-unknown-freebsd10.4
Thread model: posix
```

As per logged, I installed/built it per default/generic ways,  the FLAGS needed is in those scripts. Why I asked you TO READ THE F** LOG!!
```
sys/install.sh
```

**Comment by nsajko** [comment]
> As per logged, I installed/built it per default/generic ways, the FLAGS needed is in those scripts. Why I asked you TO READ THE F** LOG!!

You seem to be implying that the CFLAGS environment variable was unset when you called `sys/install.sh`, but it is not clear that it was so ...

Did you manually run `sys/install.sh` or was it done through some system/distro specific build system?

Can you run `echo $CFLAGS` in your shell, in the same environment from which you ran `sys/install.sh`?

**Comment by radare** [comment]
everybody calm down.

We are having so much violence lately in r2land. The increase of userbase implies more people complaining faster about build problems on random targets we don’t test.

I’m not a big fan of the restrict keyword, in fact it has been removed from the C++ standard because it is ignored by most compilers.

Everyone who contributes to r2 do this to improve the project, and sometimes we all get picky for some decissions or changes in the project.

@nsajko, is there any problem if we just remove the restrict keyword? or at least put it only for the systems we know it uses it, but i think it’s more related to version of libc. we shouldnt depend on custom cflags to make r2 build otherwise we get less portable to different or old compilers or architectures.

**Comment by nsajko** [comment]
I am not proposing some exotic CFLAGS, but https://github.com/radare/radare2/issues/11346 would maybe solve unixfreaxjp's issues.

**Comment by unixfreaxjp** [comment]
> I installed/built it per default/generic ways,

means I don't set $CFLAGS myself, and I am in purpose to make sure it is unset before building r2.
unset by means that the value of $CFLAGS is NULL. 

> Did you manually run `sys/install.sh`

As you see it in the LOG. And we are all using the git version, this is why I paste you the command I executed in shell to built r2. 

I don't think we need `restrict`, and without it we have no problem, and know we have no problem again after @radare suggest the problem vector and I struggle to remove + test it before @Maijin pullreq it.

**Comment by unixfreaxjp** [comment]
> I am not proposing some exotic CFLAGS, but #11346 would maybe solve unixfreaxjp's issues

Don't need your exotic $CFLAGS and happy to have r2 coded WITHOUT `restrict`, period.

**Comment by unixfreaxjp** [comment]
> we shouldnt depend on custom cflags to make r2 build otherwise we get less 
> portable to different or old compilers or architectures.

Hear hear! Agreed!

**Comment by nsajko** [comment]
> I’m not a big fan of the restrict keyword, in fact it has been removed from the C++ standard because it is ignored by most compilers.

I, too, think using `restrict` without a good reason in r2 would probably amount to premature optimization; but this is a case where we basically have to assume `restrict` is being used even if `restrict` is not in radare2 code, because PrintfCallback sometimes holds a pointer to `printf`, which has `restrict` in the type of the format string. Thus it would be more clear and documentary to use `restrict` in PrintfCallback, too.

BTW, `restrict` is not ignored, see https://stackoverflow.com/questions/1965487/does-the-restrict-keyword-provide-significant-benefits-in-gcc-g

**Comment by nsajko** [comment]
> FACT: Your implementation of restrict was causing problem < this is the HOW and WHY

* I did not implement `restrict`, I wish I had.

* The problem is probably that your build specified '-ansi' or '-std=c90', or something like that, but that will be fixed probably by https://github.com/radare/radare2/issues/11346

**Comment by unixfreaxjp** [comment]
> The problem is probably that your build specified '-ansi' or '-std=c90', or something like that, 

You mean my build(s)? Those were 6 systems, multiple arch and OS, and runs well before. And I also explained I do stuff generic ways since I knew if I do non-standard ppl like you will use it as excuse.

I am happy to hear more excuse you wanna say..

**Comment by ret2libc** [comment]
It's a fact that sometimes bad things happen in the git master branch. They usually last for few commits. Some times master couldn't even compile on *any* compiler. And it had happened to everybody, "core" contributors or the main maintainer included. We do our best to prevent and avoid it, but sometimes it just happens and we try to fix it ASAP.

That said, I think that people should use releases (last one is 2.9.0) to be sure everything compiles and works "well enough". Git master branch is the active development branch, so it is expected that from time to time things may break, even if we try to avoid that.

**Comment by unixfreaxjp** [comment]
> That said, I think that people should use releases (last one is 2.9.0) to be sure everything compiles and works "well enough".

I was saying the exact same tone too from a while back, agreed. 
Where I grew up, "dev" version and "released" version should always be in separate usage. And userland is advised to use "released" version.

**Comment by radare** [comment]
releases also break, last 3 releases are broken for android-mips for example, i think i fixed this but didnt had time to test yet. But as a community we should understand and improve the project to make everyone happy without .

so.. any way we can test *bsd builds in travis? seems like we need separate build machines with jenkins to me..

https://www.youtube.com/watch?v=SCLfKWaUGa8 <https://www.youtube.com/watch?v=SCLfKWaUGa8>

**Comment by XVilka** [comment]
No. but we have a good performant server with Jenkins which can run FreeBSD/whatever in virtual machine. Someone just need to take care/maintain it.

**Comment by XVilka** [comment]
The one where http://ci.rada.re is hosted.

**Comment by radare** [comment]
lol https://www.freebsdnews.com/2016/10/07/running-freebsd-travis-ci/

**Comment by XVilka** [comment]
I think running qemu in Travis will be uberslow.

**Comment by unixfreaxjp** [comment]
I can run on  x86_32 or x86_64 in daily basis (once per day) to check broken compilation. It's not Travis but that can be done right away.

**Comment by unixfreaxjp** [comment]
> No. but we have a good performant server with Jenkins which can run FreeBSD/whatever in virtual machine. 

VM with Free/OpenBSD looks nice too, but that's mostly KVM basis, or you want to do vmware is nicer.

> Someone just need to take care/maintain it.

I can help maintaining freebsd for security and updates if needed.

**Pull request merged**